<a href="https://colab.research.google.com/github/annchirackal/Tensorflow-Developer-Certification/blob/TF_Transfer_Learning/12_Transfer_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#####What is transfer learning?<br>

we can leverage an existing nueral network architecture proven to work on similar problems which is already learned patterns from dataset  similar  to our own.



Download the data:
we are going to use only 10 percent data, this would help us to understand the powe of transfer learning.

In [7]:
import tensorflow as tf
import zipfile
import os
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import datetime

In [2]:
## Global variables
IMAGE_SHAPE=(224,224)
BATCH_SIZE=32
EPOCHS=5

In [3]:
 !wget -q https://storage.googleapis.com/ztm_tf_course/food_vision/10_food_classes_10_percent.zip

In [4]:
# unzip the loaded file
zip_ref=zipfile.ZipFile("/content/10_food_classes_10_percent.zip")
zip_ref.extractall()
zip_ref.close()

In [5]:
  #lets inspect the directory
for dirpath,dirnames,filenames in os.walk("10_food_classes_10_percent"):

    print(f"{len(filenames)} files and  {len(dirnames)} in directory {dirpath}")


0 files and  2 in directory 10_food_classes_10_percent
0 files and  10 in directory 10_food_classes_10_percent/train
75 files and  0 in directory 10_food_classes_10_percent/train/chicken_curry
75 files and  0 in directory 10_food_classes_10_percent/train/pizza
75 files and  0 in directory 10_food_classes_10_percent/train/hamburger
75 files and  0 in directory 10_food_classes_10_percent/train/sushi
75 files and  0 in directory 10_food_classes_10_percent/train/chicken_wings
75 files and  0 in directory 10_food_classes_10_percent/train/grilled_salmon
75 files and  0 in directory 10_food_classes_10_percent/train/fried_rice
75 files and  0 in directory 10_food_classes_10_percent/train/steak
75 files and  0 in directory 10_food_classes_10_percent/train/ice_cream
75 files and  0 in directory 10_food_classes_10_percent/train/ramen
0 files and  10 in directory 10_food_classes_10_percent/test
250 files and  0 in directory 10_food_classes_10_percent/test/chicken_curry
250 files and  0 in director

We have 75 images of each class for traning and 250 from each class for validation

In [6]:
 #### set the path variables

train_dir="/content/10_food_classes_10_percent/train"
test_dir="/content/10_food_classes_10_percent/test"

train_data_gen=ImageDataGenerator(rescale=1./255,)
test_data_gen=ImageDataGenerator(rescale=1./255,)
print("Traning Images:")
train_data=train_data_gen.flow_from_directory(train_dir,
                                              target_size=IMAGE_SHAPE,
                                              batch_size=BATCH_SIZE,
                                              class_mode="categorical")
print("Testing Images:")
train_data=train_data_gen.flow_from_directory(test_dir,
                                              target_size=IMAGE_SHAPE,
                                              batch_size=BATCH_SIZE,
                                              class_mode="categorical")


Traning Images:
Found 750 images belonging to 10 classes.
Testing Images:
Found 2500 images belonging to 10 classes.


Callbacks are extra functionality you can add to your model to be performed during or after training.Some of the most popular call backs are
- Model check point with the ModelCheckpoint callbacks.
- Tracking experiment with tensor board callbacks.
- Stopping a model from traning with EarlyStopping callbacks.

In [10]:
#create a tensorboard callbacks

def create_tensorboard_callbacks(dirname,experiment_name):
  log_dir=dirname+"/"+experiment_name+datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
  tensorboard_callback=tf.keras.callbacks.TensorBoard(log_dir=log_dir,)
  print(f"Saving TensorBorad log files to {log_dir} ")
  return tensorboard_callback


Creating Models using Tensorflow hub<br>

In previous notebooks we have created custom models from scratch. Here we are creating a model by using layers from pretrained models availble in tensor hub.
